### ISSUE TITLE
React Components cannot be dynamically imported

### ISSUE THREAD

**Issue by fogoplayer** [codesandbox/codesandbox-client#953]
# 🐛 bug report

## Description of the problem
When I try to use dynamic import() to render a React component, I get the error `Uncaught (in promise) Error: [module url in question] hasn't been transpiled yet.` If I statically import the file first, this is no longer an issue, but that defeats the purpose of dynamic imports.

## How has this issue affected you? What are you trying to accomplish?
I am trying to dynamically import a React component.

### Your Environment
| Software         | Name/Version|
| ---------------- | ---------- |
| Сodesandbox  | 
| Browser | Chrome 6 |
| Operating System | Chrome OS |

**Comment by CompuIves** [comment]
Heya! Do you have a reproducible sandbox? I can't reproduce the issue with this one: https://codesandbox.io/s/6193vkw44k

**Comment by CompuIves** [comment]
I'm closing this one now, if there's a reproducible error I'll reopen it again!

**Comment by fogoplayer** [comment]
Sorry it took me so long to get back to you on this.

I was able to reproduce it, but only if the imported component is in a different folder than the importing component, and if the argument passed to `import()` is a variable. So 
    
    import("../components/component") //Works
    
    const componentPath = "../components/component"
    import(componentPath) // Uncaught (in promise) Error: /src/components/component.js hasn't been transpiled yet.

Here's a sandbox demonstrating the issue: https://codesandbox.io/s/4wzo1ml2o4 The relevant code is in `src/pages/import.js`.

Thank you, and sorry again.

**Comment by fogoplayer** [comment]
I've just found another situation in which code doesn't transpile. To reproduce it, set up your file system like this:
```
components
|-importingFile.js
|-Subfolder
```
In the subfolder, create an untyped file:
```
components
|-importingFile.js
|-Subfolder
  |-ImportedFile
```
And rename it to be a `.js` file:
```
components
|-importingFile.js
|-Subfolder
  |-ImportedFile.js
```
When you attempt to dynamically import ImportedFile.js in importingFile.js using a variable for the path, it will return the same error:
```
// importingFile.js
onst componentPath = "./Subfolder/ImportedFile.js"
import(componentPath) // Uncaught (in promise) Error: /src/Subfolder/ImportedFile.js hasn't been transpiled yet.
```

However, there is one other way that this differs from what was described above: once you import it once using a static import or a dynamic import with a string as its parameter, you can then switch back to the dynamic import with variable path and it will transpile correctly, even if the imported file is edited.

If this deserves a separate issue, I'll make one, but the two situations are similar enough that I figured it made more sense as a comment on this one.

**Comment by CompuIves** [comment]
Hmm, yeah computed imports with variables are a bit tricky to statically analyze. We could maybe do constant folding to replace the variable with the string using a babel plugins, but this could potentionally make transpilation much slower.

The second issue you described is indeed related, since we were able to statically analyze the initial import we were able to transpile it and use it in the next run when it was dynamic (by chance really).

The solution would be constant folding, I need to decide if we want to pay this cost though 🙂.
